### PR TITLE
ZIP 239 preparations 2

### PR DESCRIPTION
**Pull request by str4d**
Cherry-picked from the following upstream PRs:
- bitcoin/bitcoin#6722
  - Only the ancillary commits, not the mempool limiting commits (we have our own).
- bitcoin/bitcoin#6898
  - Only the first three commits (we'll cherry-pick the main content later).
- bitcoin/bitcoin#7840

### CODE REVIEW

**Comment by str4d** [comment]
Best reviewed by-commit.

@zkbot try

**Comment by str4d** [comment]
Fixed the failing RPC test.

@zkbot try

**Review by nuttycom — APPROVED** [review]
utACK

File: src/main.cpp [hunk 1]
```diff
@@ -7297,6 +7275,22 @@ bool ProcessMessages(const CChainParams& chainparams, CNode* pfrom)
...
+public:
+    CompareInvMempoolOrder(CTxMemPool *mempool)
+    {
+        mp = mempool;
+    }
+
+    bool operator()(std::set<uint256>::iterator a, std::set<uint256>::iterator b)
+    {
+        /* As std::make_heap produces a max-heap, we want the entries with the
+         * fewest ancestors/highest fee to sort later. */
```
Comment (daira) [inline src/main.cpp:7290]:
```suggestion
         * highest fee to sort later. */
```
The upstream comment doesn't reflect what we actually implement in zcashd.

File: src/main.cpp [hunk 2]
```diff
@@ -7427,49 +7421,109 @@ bool SendMessages(const Consensus::Params& params, CNode* pto)
...
+                    if (fInMemPool && IsExpiringSoonTx(tx, currentHeight + 1)) {
                         continue;
                     }
-                }
 
-                pto->filterInventoryKnown.insert(inv.hash);
+                    CInv inv(MSG_TX, hash);
+                    pto->setInventoryTxToSend.erase(hash);
+                    if (pto->pfilter) {
+                        if (!fInMemPool) continue; // another thread removed since queryHashes, maybe...
```
Comment (daira) [inline src/main.cpp:7471]: I don't understand the comment.

**Review by daira — APPROVED** [review]
utACK with comments.

File: src/main.cpp (hunk 2 above)
Comment (str4d) [reply src/main.cpp:7471]: It's saying that if we're in this scope, the transaction with txid `hash` _was_ in the mempool a few lines earlier (when `mempool.queryHashes` was called), so there was likely a race with another thread that caused it to be removed from the mempool.

**Comment by str4d** [comment]
@zkbot r+

**Review by daira — APPROVED** [review]
utACK fe7abcad1423f1a320453c5439b5b6b36f0d265a

**Comment by str4d** [comment]
Transient CI failure. I've bumped the timeout to hopefully account for CI being slower.

@zkbot r+

**Pull request merged**